File: lessons/03-subscribers/python/smartphone.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.msg import String

class SmartPhone(Node):
    def __init__(self):
        super().__init__("smartphone")#initialze node with name: smartphone
        #create a subscriber with (topic type, topic name, callback function, queue size)
        self.subscriber_=self.create_subscription(String,
                "robot_news",self.callback_robot_news,10)
        self.get_logger().info("Smartphone has been started")#print startup message to the console
        
    def callback_robot_news(self,msg:String):
        self.get_logger().info(f"Smartphone received: {msg.data}")#Display received message 



def main(args=None):
    rclpy.init(args=args)#initialze ROS2 communication
    node=SmartPhone()#create instance of smartphone node
    rclpy.spin(node)#keep the node alive to process incoming messages
    rclpy.shutdown()#shutdown ROS2 communciation

if __name__=="__main__":
    main()

File: lessons/03-subscribers/cpp/smartphone.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/string.hpp"

class Smartphone: public rclcpp::Node {
    public:
        Smartphone():Node("smartphone")
        {
            //create a subscriber with <topic_type>(topic_name,queue_size,callback_function)
            subscriber_=this->create_subscription<example_interfaces::msg::String> (
                "robot_news",10,
                std::bind(&Smartphone::callbackRobotNews,this,std::placeholders::_1));
            RCLCPP_INFO(this->get_logger(),"Smartphone has been started");//print startup message to the console
        }


    private:
        void callbackRobotNews(const example_interfaces::msg::String::SharedPtr msg)
        {
            RCLCPP_INFO(this->get_logger(),"Received news: %s",msg->data.c_str());
        }
        rclcpp::Subscription<example_interfaces::msg::String>::SharedPtr subscriber_;//shared pointe to subscriber object
};


int main(int argc, char **argv) {
    rclcpp::init(argc,argv);//initialize ROS2 communication
    auto node = std::make_shared<Smartphone>();//create an instance of the Smartphone node
    rclcpp::spin(node);//keep the node alive and processing incoming messages
    rclcpp::shutdown();//shutdown ROS2 communication
    return 0;
}